runtime/src/swam/runtime/Interface.scala

```diff
@@ -206,3 +206,53 @@ trait Memory[F[_]] extends Interface[F, MemType] {
     */
   def writeBytes(idx: Int, bytes: ByteBuffer): F[Unit]
 }
+
+
+import scala.concurrent._
+import java.util.concurrent.locks._
+
+import config._
+
+/** Tracers must implement this interface.
+  * Also, tracers must provide and implementation of innerTrace method
+  * Every trace event will be recorded in background, with the correct invocation order 
+  */
+
+trait Tracer{
+
+    //  Events must be written in order
+    val locker = new ReentrantLock
+    protected val conf: EngineConfiguration
+    val regex = conf.tracer.filter.r
+
+    // Relative time
+    val now = System.nanoTime()
+  
+   /** Public method to record an event
+   
+    *  $boundaries
+    */
+    def traceEvent(eventName: String, args: Any*): Unit
+  
+  
+    /** Make a background request on the event record call
+   
+    *  $boundaries
+    */
+    protected def executeOnBack(f: () => Unit) = {
```

Comment:
Hmm that one is tricky to explain in a review, I will make a suggestion using the cats-effect stuffs instead.